crates/task-impls/src/transactions.rs

```diff
@@ -312,93 +327,174 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, Ver: StaticVersionType>
         view_number: TYPES::Time,
         parent_comm_sig: &<<TYPES as NodeType>::SignatureKey as SignatureKey>::PureAssembledSignatureType,
     ) -> anyhow::Result<BuilderResponses<TYPES>> {
-        let available_blocks = self
-            .builder_client
-            .available_blocks(
-                parent_comm,
-                view_number.u64(),
-                self.public_key.clone(),
-                parent_comm_sig,
-            )
-            .await
-            .context("getting available blocks")?;
-        tracing::debug!("Got available blocks: {available_blocks:?}");
-
-        let block_info = available_blocks
-            .into_iter()
-            .max_by(|l, r| {
-                // We want the block with the highest fee per byte of data we're going to have to
-                // process, thus our comparision function is:
-                //      (l.offered_fee / l.block_size) < (r.offered_fee / r.block_size)
-                // To avoid floating point math (which doesn't even have an `Ord` impl) we multiply
-                // through by the denominators to get
-                //      l.offered_fee * r.block_size < r.offered_fee * l.block_size
-                // We cast up to u128 to avoid overflow.
-                (u128::from(l.offered_fee) * u128::from(r.block_size))
-                    .cmp(&(u128::from(r.offered_fee) * u128::from(l.block_size)))
-            })
-            .context("no available blocks")?;
-        tracing::debug!("Selected block: {block_info:?}");
-
-        // Verify signature over chosen block.
-        if !block_info.sender.validate_block_info_signature(
-            &block_info.signature,
-            block_info.block_size,
-            block_info.offered_fee,
-            &block_info.block_hash,
-        ) {
-            bail!("Failed to verify available block info response message signature");
-        }
+        // Create a future for every builder
```

Comment:
This function is quite long, is it possible to pare it down some? It looks like we're doing a lot of logic that's not necessarily tightly related.